File: Task3C/stockitem.cpp
#include "stockitem.h"

// Default constructor
StockItem::StockItem() {
    this->product = NULL;
    this->quantity = 0;
}

// Parameterised constructor
StockItem::StockItem(Product* product, int quantity) {
    this->product = product;
    this->quantity = quantity;
}

// Getters and setters
Product* StockItem::getProduct() {
    return this->product;
}

int StockItem::getQuantity() {
    return this->quantity;
}

void StockItem::setProduct(Product* product) {
    this->product = product;
}

void StockItem::setQuantity(int quantity) {
    this->quantity = quantity;
}

// Save stock item information
void StockItem::save(ostream& out) const {
    this->product->save(out);
    out << quantity << "\n";
}

// Load stock item information
void StockItem::load(istream& in) {
    product->load(in);
    
    string line;
    getline(in, line);
    stringstream ss(line);

    string token;
    getline(ss, token);
    quantity = stoi(token);
    
}

// Destructor
StockItem::~StockItem() {
    delete this->product;
}

File: Task2A_B/task2a_b/src/main/java/com/cps2004/StockItem.java
package com.cps2004;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

public class StockItem {
    private Product product;
    private int quantity;

    // Default constructor
    public StockItem(){
        this.product = null;
        this.quantity = 0;
    }

    // Parameterized constructor
    public StockItem(Product product, int quantity){
        this.product = product;
        this.quantity = quantity;
    }

    // Getters and setters
    public Product getProduct() { return product; }
    public int getQuantity() { return quantity; }
    public void setProduct(Product product) { this.product = product; }
    public void setQuantity(int quantity) { this.quantity = quantity; }

    // Save product information to file
    public void save(ObjectNode newItem) {
        product.save(newItem);
        newItem.put("quantity", this.quantity);
    }

    // Load product information from file
    public void load(JsonNode node) {
        product.load(node);
        quantity = node.get("quantity").asInt();
    }
}
